lapack/gonum/dtgsy2.go

```diff
@@ -0,0 +1,696 @@
+					}
+					if j < q {
+						bi.Dger(mb, n-je-1, 1, rhs[2:], 1, b[js*ldb+je+1:], 1,
+							c[is*ldc+je+1:], ldc)
+						bi.Dger(mb, n-je-1, 1, rhs[2:], 1, e[js*lde+je+1:], 1,
+							f[is*ldf+je+1:], ldf)
+					}
+				} else if mb == 2 && nb == 2 {
+					// Build 8×8 system Z * x = rhs
+					impl.Dlaset(blas.All, ldz, ldz, 0, 0, z, ldz)
+
+					z[0] = a[is*lda+is]
+					z[ldz] = a[isp1*lda+is]
+					z[4*ldz] = d[is*ldd+is]
+
+					z[1] = a[is*lda+isp1]
+					z[ldz+1] = a[isp1*lda+isp1]
+					z[4*ldz+1] = d[is*ldd+isp1]
+					z[5*ldz+1] = d[isp1*ldd+isp1]
+
+					z[2*ldz+2] = a[is*lda+is]
+					z[3*ldz+2] = a[isp1*lda+is]
+					z[6*ldz+2] = d[is*ldd+is]
+
+					z[2*ldz+3] = a[is*lda+isp1]
+					z[3*ldz+3] = a[isp1*lda+isp1]
+					z[6*ldz+3] = d[is*ldd+isp1]
+					z[7*ldz+3] = d[isp1*ldd+isp1]
+
+					z[4] = -b[js*ldb+js]
+					z[2*ldz+4] = -b[js*ldb+jsp1]
+					z[4*ldz+4] = -e[js*lde+js]
+					z[6*ldz+4] = -e[js*lde+jsp1]
+
+					z[ldz+5] = -b[js*ldb+js]
+					z[3*ldz+5] = -b[js*ldb+jsp1]
+					z[5*ldz+5] = -e[js*lde+js]
+					z[7*ldz+5] = -e[js*lde+jsp1]
+
+					z[6] = -b[jsp1*ldb+js]
+					z[2*ldz+6] = -b[jsp1*ldb+jsp1]
+					z[6*ldz+6] = -e[jsp1*lde+jsp1]
+
+					z[ldz+7] = -b[jsp1*ldb+js]
+					z[3*ldz+7] = -b[jsp1*ldb+jsp1]
+					z[7*ldz+7] = -e[jsp1*lde+jsp1]
+
+					// Set up right hand side(s).
+					k = 0
+					ii := mb * nb
+					for jj := 0; jj <= nb-1; jj++ {
+						bi.Dcopy(mb, c[is*ldc+js+jj:], ldc, rhs[k:], 1)
+						bi.Dcopy(mb, f[is*ldf+js+jj:], ldf, rhs[ii:], 1)
+						k += mb
+						ii += mb
+					}
+
+					// Solve Z * x = rhs
+					k = impl.Dgetc2(zdim, z, ldz, ipiv[:zdim], jpiv[:zdim])
+					if k >= 0 {
+						info = k
+					}
+					if ijob == 0 {
+						scaloc := impl.Dgesc2(zdim, z, ldz, rhs, ipiv[:zdim], jpiv[:zdim])
+						if scaloc != 1 {
+							for k = 0; k < n; k++ {
+								bi.Dscal(m, scaloc, c[k:], ldc)
+								bi.Dscal(m, scaloc, f[k:], ldf)
+							}
+							scale *= scaloc
+						}
+					} else {
+						rdsum, rdscal = impl.Dlatdf(ijob, zdim, z, ldz, rhs,
+							rdsum, rdscal, ipiv[:zdim], jpiv[:zdim])
+					}
+
+					// Unpack solution vectors(s).
+					k = 0
+					ii = mb * nb
+					for jj := 0; jj <= nb-1; jj++ {
+						bi.Dcopy(mb, rhs[k:], 1, c[is*ldc+js+jj:], ldc)
+						bi.Dcopy(mb, rhs[ii:], 1, f[is*ldf+js+jj:], ldf)
+						k += mb
+						ii += mb
+					}
+
+					// Substitute R(i, j) and L(i, j) into remaining equation.
+					if i > 0 {
+						bi.Dgemm(blas.NoTrans, blas.Trans, is, nb, mb, -1,
+							a[is:], lda, rhs, mb, 1, c[js:], ldc)
+						bi.Dgemm(blas.NoTrans, blas.Trans, is, nb, mb, -1,
+							d[is:], ldd, rhs, mb, 1, f[js:], ldf)
+					}
+					if j < q {
+						k = mb * nb
+						bi.Dgemm(blas.Trans, blas.NoTrans, mb, n-je-1, nb, 1,
+							rhs[k:], mb, b[js*ldb+je+1:], ldb, 1, c[is*ldc+je+1:], ldc)
+						bi.Dgemm(blas.Trans, blas.NoTrans, mb, n-je-1, nb, 1,
+							rhs[k:], mb, e[js*lde+je+1:], lde, 1, f[is*ldf+je+1:], ldf)
```

Comment:
@vladimir-ch I've found the issue lies with gonum's Dgemm implementation which can't emulate the full functionality of Fortran's Dgemm for non-transposed matrices due to the row-major and column major difference. If there were a column major Dgemm in gonum I could easily fix this PR.

Reply:
https://groups.google.com/g/gonum-dev/c/AJFZHgTaxZ8/m/7ahKK2SCAQAJ

Reply:
This is not an issue with gonum's Dgemm. It does exactly what the reference implementation does, only the matrices are represented differently. The issue is that Dtgsy2 does not adapt to this different representation correctly.

Besides, the test fails even when m=1,2, n=1,2 which doesn't have Dgemm in its code path, so there must be other issues as well.

Reply:
So oddly I've managed to have Dtgsy2 passing for cases with practically any n,m combinations using `NoTrans` but when `ijob` is zero. If ijob is not zero a code path with Dlatdf at lines ~188, ~249, ~314, ~400 is taken.